calyx/src/backend/verilog.rs

```diff
@@ -370,22 +371,24 @@ fn memory_read_write(comp: &ir::Component) -> Vec<v::Stmt> {
             ],
         )));
 
-    let memories = comp.cells.iter().filter_map(|cell| {
-        let is_external = cell.borrow().get_attribute("external").is_some();
-        if is_external
-            && cell
-                .borrow()
-                .type_name()
-                .map(|proto| proto.id.contains("mem"))
-                .unwrap_or_default()
-        {
-            Some(cell.borrow().name.id.clone())
-        } else {
-            None
-        }
-    });
+    let memories = || {
```

Comment:
Why is it necessary to make this a closure?

Reply:
This is a bit funky. The short version is that the return type of `iter_cells` and `iter_groups` is a _thing_ that implements Iterator returning references to RRCs of the appropriate type. I figured this was a reasonable encapsulation since it wasn't necessary to expose precisely how the iterator was constructed. Unfortunately, this has the side-effect of _only_ providing functionality related to the `Iterator` trait and in this particular case, it means that the iterators are not cloneable. Since this code wanted to construct an iterator and then clone it (rather than construct it twice) it ran into an error when trying to clone it. Rather than construct the iterator multiple times explicitly, I opted for a smaller code change to make what was originally the iterator a closure that returns an iterator. I would imagine this is performance-wise the same as the code that clones it.

If this seems overly silly, a possible solution would be to create a custom Trait that requires both Iterator and Clone and return something of that type when iterating over the cells and groups rather than returning something that only promises to implement Iterator.

Reply:
I'm in favor of creating a custom iterator rather than having those methods return `impl Iterator<..>`. And then we can make this thing cloneable and it probably gives us more flexibility in the future

Reply:
Do you mean making a specific Iterator implementation or making a "super"-trait to encompass both `Clone` and `Iterator`?

If the former, we can just change the return type to the concrete iterator being returned, rather than "something that implements Iterator". Conceptually, I don't really have a problem with it, but it does feel like it is a bit of a leaky API since things outside shouldn't have to care about what exact thing is implementing iterator

Reply:
Yeah I see what you mean. But this also how a lot of Rust libraries seem to work? I think? Not entirely sure why tbh

Reply:
Seems a bit odd to me, but I guess it's not a big deal as long as people don't end up depending on special features of a given iterator implementation which aren't necessarily guaranteed

Reply:
Turns out `impl Clone + Iterator<Item = &RRC<T>>` is totally allowed. I feel very silly.